File: week12/examples/lab12/p_c_demo3.cpp
#include <iostream>
#include <cstring>
using namespace std;
class Parent{
private:
    int id;
    char* name;

public:
    Parent(int i=0, const char* n="null");
    Parent(const Parent& p);
    virtual ~Parent();
    Parent& operator=(const Parent& prhs);

    friend ostream& operator<<(ostream& os, const Parent& p){
        os<<"Parent:"<<p.id<<", "<<p.name<<endl;
        return os;
    } 
};

class Child: public Parent{
private:
    char* style;
    int age; 

public:
    Child(int i=0, const char* n="null", const char* s="null", int a=0);
    Child(const Child& c);
    ~Child();
    Child& operator=(const Child& crhs);
    friend ostream& operator<<(ostream& os, const Child& c){
        os<<(Parent&)c<<"Child:"<<c.style<<", "<<c.age<<endl;
        return os;
    } 
};

Parent::Parent(int i, const char* n){
    cout<<"calling Parent defautl constructor Parent()\n";
    id = i;
    name = new char[strlen(n) + 1];
    //strcpy_s(name,strlen(n)+1, n);
    strncpy(name, n,strlen(n)+1);
}

Child::Child(int i, const char* n, const char* s, int a): Parent(i,n){
    cout<<"call Child default constructor Child()\n";
    style = new char[strlen(s) + 1];
    //strcpy_s(style,strlen(s)+1, s);
    strncpy(style, s,strlen(s)+1);
    age=a;
}

Parent:: ~Parent(){
    cout<< "call Parent destructor.\n";
    delete [] name;
}
Child::~Child(){
    cout<<"call Child destructor.\n";
    delete[] style;
}

Parent::Parent(const Parent& p){
    cout<<"calling Parent copy constructor Parent(const Parent&)\n";
    id = p.id;
    name = new char[strlen(p.name)+1];
    //strcpy_s(name, strlen(p.name)+1, p.name);
    strncpy(name, p.name,strlen(p.name)+1);
}

Child::Child(const Child& c):Parent(c){
    cout<<"calling Child copy constructor Child(const Child&)\n";
    age = c.age;
    style = new char[strlen(c.style)+1];
    //strcpy_s(style, strlen(c.style)+1, c.style);
    strncpy(style, c.style, strlen(c.style)+1);
}

Parent& Parent::operator=(const Parent& prhs){
    cout<<"call Parent assignment operator:\n";
    if(this == &prhs)
        return *this;

    delete []name;
    this->id = prhs.id;
    name = new char[strlen(prhs.name)+1];
    //strcpy_s(name,strlen(prhs.name)+1, prhs.name);
    strncpy(name,prhs.name,strlen(prhs.name)+1);
    return *this;    
}

Child& Child::operator=(const Child& crhs){
    cout<<"call Child assignment operator:\n";
    if(this == &crhs)
        return *this;
    Parent::operator=(crhs);

    delete []style;
    style = new char[strlen(crhs.style)+1];
    //strcpy_s(style,strlen(crhs.style)+1,crhs.style);
    strncpy(style,crhs.style,strlen(crhs.style)+1);
    age = crhs.age;

    return *this;    
}

int main(){
    Parent p1;
    cout<< "value in p1\n"<<p1<<endl;

    Parent p2(101, "Liming");
    cout<< "value in p2\n"<<p2<<endl;

    Parent p3(p1);
    cout<< "value in p3\n"<<p3<<endl;
    p1 = p2;
    cout<< "value in p1\n"<<p1<<endl;

    Child c1;
    cout<<"value in c1\n"<<c1<<endl;

    Child c2(201, "Wuhong","teenager",15);
    cout<<"value in c2\n"<<c2<<endl;

    Child c3(c1);
    cout<< "value in c3\n"<<c3<<endl;
    c1=c2;
    cout<<"value in c1\n"<<c1<<endl;

    return 0;
}

File: week12/examples/lab12/p_c_demo3.py
import copy

class Parent:
    def __init__(self, i=0, n="null"):
        print("calling Parent default constructor Parent()")
        self.id = i
        self.name = n  # Python 字符串不可变，无需手动内存管理

    def __deepcopy__(self, memo):
        print("calling Parent copy constructor Parent(const Parent&)")
        new_obj = self.__class__(self.id, self.name)
        memo[id(self)] = new_obj
        return new_obj

    def assign(self, other):
        print("call Parent assignment operator:")
        if self is other:
            return self
        self.id = other.id
        self.name = other.name
        return self

    def __del__(self):
        print("call Parent destructor.")  # 实际无需手动释放资源

    def __str__(self):
        return f"Parent:{self.id}, {self.name}\n"

class Child(Parent):
    def __init__(self, i=0, n="null", s="null", a=0):
        super().__init__(i, n)
        print("call Child default constructor Child()")
        self.style = s
        self.age = a

    def __deepcopy__(self, memo):
        print("calling Child copy constructor Child(const Child&)")
        new_obj = self.__class__(self.id, self.name, self.style, self.age)
        memo[id(self)] = new_obj
        return new_obj

    def assign(self, other):
        print("call Child assignment operator:")
        if self is other:
            return self
        super().assign(other)
        self.style = other.style
        self.age = other.age
        return self

    def __del__(self):
        super().__del__()
        print("call Child destructor.")  # 父类 __del__ 会自动调用

    def __str__(self):
        #parent_str = super().__str__().replace("Parent:", "Child:", 1)
        parent_str = super().__str__()
        return f"{parent_str}Child:{self.style}, {self.age}\n"

if __name__ == "__main__":
    # 模拟 C++ 对象构造
    p1 = Parent()
    print("value in p1\n", p1)

    p2 = Parent(101, "Liming")
    print("value in p2\n", p2)

    p3 = copy.deepcopy(p1)
    print("value in p3\n", p3)

    p1.assign(p2)
    print("value in p1\n", p1)

    # 子类测试
    c1 = Child()
    print("value in c1\n", c1)

    c2 = Child(201, "Wuhong", "teenager", 15)
    print("value in c2\n", c2)

    c3 = copy.deepcopy(c1)
    print("value in c3\n", c3)

    c1.assign(c2)
    print("value in c1\n", c1)

    # 手动触发析构（Python 垃圾回收时机不确定）
    del  c3, c2, c1, p3, p2, p1

